### PR TITLE
Expose and scrape metrics from `machine-controller-manager-provider-xxx`

### PR DESCRIPTION
**Pull request by aaronfern**
**How to categorize this PR?**

/area monitoring
/kind enhancement

**What this PR does / why we need it**:
This PR does the following
- Updates the `MCM` deployment and service so that metrics from the `machine-controller-manager-provider-xxx` container are exposed
- Update the `serviceMonitor` so that prometheus can scrape these metrics

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```other operator
`machine-controller-manager-provider-xxx` container now exposes metrics that prometheus can scrapes
```

### CODE REVIEW

**Review by rfranzke — COMMENTED** [review]
Technically, the better place to mutate `Service` and `ServiceMonitor` for the MCM deployment would be in the provider extensions themselves, for example https://github.com/gardener/gardener/blob/master/pkg/provider-local/webhook/controlplane/ensurer.go (like [this](https://github.com/gardener/gardener/blob/master/pkg/provider-local/webhook/controlplane/ensurer.go#L41-L72)), since `gardener/gardener` does not know about the provider sidecar containers.

However, given that this seems to always be the same for all providers, we can pragmatically decide to follow this proposed approach, such that we don't have to duplicate/replicate it in all provider repos.

@aaronfern Is this change compatible with older provider sidecars? Before this PR, the metrics were scraped from the main MCM container.

File: pkg/component/nodemanagement/machinecontrollermanager/machine_controller_manager.go [hunk 1]
```diff
@@ -320,78 +333,89 @@ func (m *machineControllerManager) Deploy(ctx context.Context) error {
...
+				},
+				{
+					Port: portNameProviderMetrics,
+					RelabelConfigs: []monitoringv1.RelabelConfig{{
+						Action: "labelmap",
+						Regex:  `__meta_kubernetes_service_label_(.+)`,
+					}},
+					MetricRelabelConfigs: monitoringutils.StandardMetricRelabelConfig(
+						// Machine related metrics
+						"mcm_machine_stale_machines_total",
```
Comment (rishabh-11) [inline pkg/component/nodemanagement/machinecontrollermanager/machine_controller_manager.go]: `"mcm_machine_stale_machines_total"` is an MCM metric

File: pkg/component/nodemanagement/machinecontrollermanager/machine_controller_manager.go [hunk 2]
```diff
@@ -320,78 +333,89 @@ func (m *machineControllerManager) Deploy(ctx context.Context) error {
...
+						"mcm_machine_set_info_spec_replicas",
+						"mcm_machine_set_info_spec_min_ready_seconds",
+						"mcm_machine_set_status_condition",
+						"mcm_machine_set_status_available_replicas",
+						"mcm_machine_set_status_fully_labelled_replicas",
+						"mcm_machine_set_status_replicas",
+						"mcm_machine_set_status_ready_replicas",
+						// misc metrics
+						"mcm_misc_scrape_failure_total",
+						"mcm_machine_controller_frozen",
```
Comment (rishabh-11) [inline pkg/component/nodemanagement/machinecontrollermanager/machine_controller_manager.go]: `"mcm_machine_controller_frozen"` is part of mc metrics

File: pkg/component/nodemanagement/machinecontrollermanager/machine_controller_manager.go [hunk 3]
```diff
@@ -320,78 +333,89 @@ func (m *machineControllerManager) Deploy(ctx context.Context) error {
...
+						"mcm_workqueue_depth",
+						"mcm_workqueue_queue_duration_seconds_bucket",
+						"mcm_workqueue_queue_duration_seconds_sum",
+						"mcm_workqueue_queue_duration_seconds_count",
+						"mcm_workqueue_work_duration_seconds_bucket",
+						"mcm_workqueue_work_duration_seconds_sum",
+						"mcm_workqueue_work_duration_seconds_count",
+						"mcm_workqueue_unfinished_work_seconds",
+						"mcm_workqueue_longest_running_processor_seconds",
+						"mcm_workqueue_retries_total",
```
Comment (rishabh-11) [inline pkg/component/nodemanagement/machinecontrollermanager/machine_controller_manager.go:391]: These should be common for both containers

**Review by rishabh-11 — CHANGES_REQUESTED** [review]
some changes needed

**Comment by aaronfern** [comment]
> @aaronfern Is this change compatible with older provider sidecars? Before this PR, the metrics were scraped from the main MCM container.

Yes, I don't see why not. There's really no change to the provider sidecar logic. Provider metrics were always exposed there, just never being scraped.

**Review by rishabh-11 — COMMENTED** [review]
/lgtm

File: pkg/component/nodemanagement/machinecontrollermanager/machine_controller_manager.go [hunk 4]
```diff
@@ -320,78 +333,105 @@ func (m *machineControllerManager) Deploy(ctx context.Context) error {
...
+					RelabelConfigs: []monitoringv1.RelabelConfig{{
+						Action: "labelmap",
+						Regex:  `__meta_kubernetes_service_label_(.+)`,
+					}},
+					MetricRelabelConfigs: monitoringutils.StandardMetricRelabelConfig(
+						// Machine related metrics
+						"mcm_machine_items_total",
+						"mcm_machine_current_status_phase",
+						"mcm_machine_info",
+						"mcm_machine_status_condition",
```
Comment (timuthy) [inline pkg/component/nodemanagement/machinecontrollermanager/machine_controller_manager.go:405]: Why have the machine related metrics moved to this place? I compared both for an AWS cluster and they seem considerably different. Is this intended?

Example for `mcm_machine_status_condition` metric in cluster with 4 machines:

MCM container:
```
# HELP mcm_machine_status_condition Information of the mcm managed Machines' status conditions.
# TYPE mcm_machine_status_condition gauge
mcm_machine_status_condition{condition="ClusterNetworkProblem",name="shoot--foo--bar-p1-z1-5c65d-j25cp",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="ClusterNetworkProblem",name="shoot--foo--bar-p2-z1-67497-hb2tk",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="ClusterNetworkProblem",name="shoot--foo--bar-p2-z2-86867-wzpg6",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="ClusterNetworkProblem",name="shoot--foo--bar-p3-z1-6b74f-fs7fd",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="CorruptDockerOverlay2",name="shoot--foo--bar-p1-z1-5c65d-j25cp",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="CorruptDockerOverlay2",name="shoot--foo--bar-p2-z1-67497-hb2tk",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="CorruptDockerOverlay2",name="shoot--foo--bar-p2-z2-86867-wzpg6",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="CorruptDockerOverlay2",name="shoot--foo--bar-p3-z1-6b74f-fs7fd",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="DiskPressure",name="shoot--foo--bar-p1-z1-5c65d-j25cp",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="DiskPressure",name="shoot--foo--bar-p2-z1-67497-hb2tk",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="DiskPressure",name="shoot--foo--bar-p2-z2-86867-wzpg6",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="DiskPressure",name="shoot--foo--bar-p3-z1-6b74f-fs7fd",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="FrequentContainerdRestart",name="shoot--foo--bar-p1-z1-5c65d-j25cp",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="FrequentContainerdRestart",name="shoot--foo--bar-p2-z1-67497-hb2tk",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="FrequentContainerdRestart",name="shoot--foo--bar-p2-z2-86867-wzpg6",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="FrequentContainerdRestart",name="shoot--foo--bar-p3-z1-6b74f-fs7fd",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="FrequentDockerRestart",name="shoot--foo--bar-p1-z1-5c65d-j25cp",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="FrequentDockerRestart",name="shoot--foo--bar-p2-z1-67497-hb2tk",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="FrequentDockerRestart",name="shoot--foo--bar-p2-z2-86867-wzpg6",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="FrequentDockerRestart",name="shoot--foo--bar-p3-z1-6b74f-fs7fd",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="FrequentKubeletRestart",name="shoot--foo--bar-p1-z1-5c65d-j25cp",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="FrequentKubeletRestart",name="shoot--foo--bar-p2-z1-67497-hb2tk",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="FrequentKubeletRestart",name="shoot--foo--bar-p2-z2-86867-wzpg6",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="FrequentKubeletRestart",name="shoot--foo--bar-p3-z1-6b74f-fs7fd",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="FrequentUnregisterNetDevice",name="shoot--foo--bar-p1-z1-5c65d-j25cp",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="FrequentUnregisterNetDevice",name="shoot--foo--bar-p2-z1-67497-hb2tk",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="FrequentUnregisterNetDevice",name="shoot--foo--bar-p2-z2-86867-wzpg6",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="FrequentUnregisterNetDevice",name="shoot--foo--bar-p3-z1-6b74f-fs7fd",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="HostNetworkProblem",name="shoot--foo--bar-p1-z1-5c65d-j25cp",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="HostNetworkProblem",name="shoot--foo--bar-p2-z1-67497-hb2tk",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="HostNetworkProblem",name="shoot--foo--bar-p2-z2-86867-wzpg6",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="HostNetworkProblem",name="shoot--foo--bar-p3-z1-6b74f-fs7fd",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="KernelDeadlock",name="shoot--foo--bar-p1-z1-5c65d-j25cp",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="KernelDeadlock",name="shoot--foo--bar-p2-z1-67497-hb2tk",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="KernelDeadlock",name="shoot--foo--bar-p2-z2-86867-wzpg6",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="KernelDeadlock",name="shoot--foo--bar-p3-z1-6b74f-fs7fd",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="MemoryPressure",name="shoot--foo--bar-p1-z1-5c65d-j25cp",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="MemoryPressure",name="shoot--foo--bar-p2-z1-67497-hb2tk",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="MemoryPressure",name="shoot--foo--bar-p2-z2-86867-wzpg6",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="MemoryPressure",name="shoot--foo--bar-p3-z1-6b74f-fs7fd",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="PIDPressure",name="shoot--foo--bar-p1-z1-5c65d-j25cp",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="PIDPressure",name="shoot--foo--bar-p2-z1-67497-hb2tk",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="PIDPressure",name="shoot--foo--bar-p2-z2-86867-wzpg6",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="PIDPressure",name="shoot--foo--bar-p3-z1-6b74f-fs7fd",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="ReadonlyFilesystem",name="shoot--foo--bar-p1-z1-5c65d-j25cp",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="ReadonlyFilesystem",name="shoot--foo--bar-p2-z1-67497-hb2tk",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="ReadonlyFilesystem",name="shoot--foo--bar-p2-z2-86867-wzpg6",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="ReadonlyFilesystem",name="shoot--foo--bar-p3-z1-6b74f-fs7fd",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="Ready",name="shoot--foo--bar-p1-z1-5c65d-j25cp",namespace="shoot--foo--bar"} 1
mcm_machine_status_condition{condition="Ready",name="shoot--foo--bar-p2-z1-67497-hb2tk",namespace="shoot--foo--bar"} 1
mcm_machine_status_condition{condition="Ready",name="shoot--foo--bar-p2-z2-86867-wzpg6",namespace="shoot--foo--bar"} 1
mcm_machine_status_condition{condition="Ready",name="shoot--foo--bar-p3-z1-6b74f-fs7fd",namespace="shoot--foo--bar"} 1
```

Provider container:
```
# HELP mcm_machine_status_condition Information of the mcm managed Machines' status conditions.
# TYPE mcm_machine_status_condition gauge
mcm_machine_status_condition{condition="ClusterNetworkProblem",name="shoot--foo--bar-p1-z1-5c65d-j25cp",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="FrequentContainerdRestart",name="shoot--foo--bar-p1-z1-5c65d-j25cp",namespace="shoot--foo--bar"} 0
mcm_machine_status_condition{condition="HostNetworkProblem",name="shoot--foo--bar-p1-z1-5c65d-j25cp",namespace="shoot--scalesim--scenario-c"} 0
```

Comment (timuthy) [reply pkg/component/nodemanagement/machinecontrollermanager/machine_controller_manager.go:405]: As confirmed by @aaronfern, machine related metrics were always exclusively exposed by the provider container. We took more samples to prove that, the one above is faulty and can be ignored.

**Review by timuthy — COMMENTED** [review]
Thanks @aaronfern 👏 
/lgtm
/approve

**Pull request merged**